index.bs

```diff
@@ -2055,86 +2055,98 @@ for the APIs that they design.
 
 <h4 id="algorithms">Defining algorithms in specifications</h4>
 
-Specification text is often more precise
-when it is written using an explicit sequence of steps.
-Using an ordered sequence of steps makes it clear how the required behaviors interact.
-For example, if there are multiple error conditions present,
-a specification that describes the behavior using a sequence of steps
-will make it clear which error will be reported,
-since the different errors would be checked and reported in different steps.
-
-However, sequences of steps
-are not the only way to make a specification precise enough.
-When there are particular characteristics
-that specification editors want to ensure the specification always satisfies,
-it may be better to use a less algorithmic way of formal specification
-that ensures these characteristics are always true.
-Developing these other ways of specifying behavior formally
-is more important when the algorithms would be more repetitive.
-These formal specification techniques can include dependencies on
-languages such as [[WEBIDL]], on grammar productions,
-or the development of new languages for defining pieces of the specification.
-Merely having a formal syntax is not sufficient, though;
-the requirements that the formal definitions place on implementations must still be clearly defined,
-which generally requires specifying some algorithms (or, sometimes, another layer of formal syntax) that define what the formal syntax means,
-and then using the more-readable formal syntax to avoid
-excessive repetition of the less-readable algorithms.
-There are also possibilities in-between, for example,
-if it is desirable to ensure that an algorithm can be implemented with a state machine,
-it may be preferable to specify it in an algorithmic way that incorporates explicit states
-so that the specification authors avoid accidentally creating more states than intended.
+Write algorithms in a way that is clear and concise.
+
+The most common way to write algorithms
+is to write an explicit sequence of steps.
+This often looks like pseudo-code.
+
+<div class="example">
+The {{HTMLDialogElement/showModal()}} method is described as a numbered sequence of steps
+which clearly explains when to throw exceptions
+and when to run algorithms defined in other parts of the HTML spec.
+</div>
+
+When writing a sequence of steps,
+imagine that it is a piece of functional code.
+
+- Clearly specify the inputs and outputs,
+name the algorithm and the variables it uses well,
+and explicitly note the points in the algorithm where the algorithm may
+return a result or error.
+- As much as possible, avoid writing algorithms which have side effects.
+
+Summarize the purpose of the algorithm before going into detail,
+so that readers can decide whether to read the steps or skip over them.
+For example take the following steps, which ensure that there is at most one pending X
+callback per toplevel browsing context".
```

Comment:
Missing opening quote

Reply:
Also, maybe "top-level" instead of toplevel.